accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -263,6 +270,113 @@ The **flatten** function is modified as follows:
+The function **futureValueType** is defined as follows:
+
+- **futureValueType**(`S?`) = **futureValueType**(`S`), for all `S`.
+- **futureValueType**(`S*`) = **futureValueType**(`S`), for all `S`.
+- **futureValueType**(`Future<S>`) = `S`, for all `S`.
+- **futureValueType**(`FutureOr<S>`) = `S`, for all `S`.
+- **futureValueType**(`void`) = `void`.
+- Otherwise, for all `S`, **futureValueType**(`S`) = `Object?`.
+
+_Note that it is a compile-time error unless the return type of an asynchronous
+non-generator function is a supertype of `Future<Never>`, which means that
+the last case will only be applied when `S` is a non-`void` top type._
+
+### Return statements
+
+The static analysis of return statements is changed in the following
+way, where `$T$` is the declared return type and `$S$` is the static type of
+the expression `e`.
+
+At [this location](https://github.com/dart-lang/language/blob/65b8267be0ebb9b3f0849e2061e6132021a4827d/specification/dartLangSpec.tex#L15477)
+about synchronous non-generator functions, the text is changed as follows:
+
+```
+It is a compile-time error if $s$ is \code{\RETURN{} $e$;},
+$T$ is neither \VOID{} nor \DYNAMIC,
+and $S$ is \VOID{} or \code{\VOID*}.
+```
+
+_Comparing to Dart before null-safety, this means that it is no longer allowed
+to "return void to null" in a regular function._
+
+At [this location](https://github.com/dart-lang/language/blob/65b8267be0ebb9b3f0849e2061e6132021a4827d/specification/dartLangSpec.tex#L15525)
+about an asynchronous non-generator function with future value type `$T_v$`,
+the text is changed as follows:
+
+```
+It is a compile-time error if $s$ is \code{\RETURN{} $e$;},
+$T_v$ is neither \VOID{} nor \DYNAMIC,
+and \flatten{S} is \VOID{} or \code{\VOID*}.
+```
+
+_Comparing to Dart before null-safety, this means that it is no longer allowed
+to "return void to null" in an `async` function, nor to "return a void future
+to null"._
+
+The next sentence is changed as follows:
+
+```
+It is a compile-time error if $s$ is \code{\RETURN{} $e$;},
+\flatten{S} is not \VOID,
+$S$ is not assignable to $T_v$,
+and flatten{S} is not a subtype of $T_v$.
+```
+
+_Comparing to Dart before null-safety, this means that it is now allowed
+to return a future when the future value type is a suitable future; for
+instance, we can have `return Future<int>.value(42)` in an `async` function
+with declared return type `Future<Future<int>>`. Conversely, it is no longer
+allowed to return a `Future<dynamic>` or `FutureOr<dynamic>` when the future
+value type is `Future<U>` for some `U` which is not a top type._
+
+The dynamic semantics specified at
+[this location](https://github.com/dart-lang/language/blob/65b8267be0ebb9b3f0849e2061e6132021a4827d/specification/dartLangSpec.tex#L15597)
+is changed as follows, where `$f$` is the enclosing function with declared
+return type `$T$`, and `$e$` is the returned expression with static type `$S$`:
```

Comment:
You don't mention `S` below, so eliminate it?

Reply:
Done.